### PR TITLE
Feature/remove timeout from connection

### PR DESCRIPTION
**Pull request by philippseith**
No timeout handling in connections. Timeouts are monitored by the handshake implementation and the signalr loop.
- Implementation of well behaving Connections is simplified.
- Parties can send signalr Close messages instead of just closing the underlying network connection

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -189,37 +190,66 @@ func (s *server) newConnectionHubContext(hubConn hubConnection) HubContext {
...
-		// Malformed handshake
-		return nil, err
+	select {
+	case result := <-readJSONFramesChan:
+		if result[1] != nil {
+			return request, result[1].(error)
+		}
+		rawHandshake := result[0].([][]byte)
+		_ = dbg.Log(evt, "handshake received", "msg", string(rawHandshake[0]))
+		if err := json.Unmarshal(rawHandshake[0], &request); err != nil {
```
Comment (andig) [inline server.go]: Simplify:

    err := …
    Return …, err

**Comment by andig** [comment]
What's missing to get this merged?

**Comment by philippseith** [comment]
While investigating the flaky test in this run https://github.com/philippseith/signalr/runs/4136023509?check_suite_focus=true , I saw the necessity to describe the meaning of the ClientStates. This leaded to an elimination of ClientError and simplification of WaitForState. The timeout in client_test.go does not appear with these changes (ran it in a loop 1000 times) and doesn't affect the other tests.
But now there is this race in the current build to investigate, and the problem in httpserver_test.go has changed (but is still flaky, ran in that error with looping over this test 100 times).

**Comment by philippseith** [comment]
> What's missing to get this merged?

a) I was waiting for more feedback
b) free time

**Comment by andig** [comment]
I'm sorry, didn't want to appear rude. I'm definitely having issues with current master: https://github.com/golang/go/issues/49543. As long as the races are not resolved I can't really follow up with the Go issue. For time being I'm stuck on 0.3.1 which doesn't have this problem.

**Comment by andig** [comment]
For what it's worth: I'm not seeing any races with this PR. Unfortunately I'm still seeing runtime crashes, but those might be unrelated. In any case they are also on master. I'll keep investigating.

**Comment by andig** [comment]
Nvm regarding the races. Looks like this is a bug in MacOS Monterey when the race detector is enabled. So this PR is working fine afaic.

**Pull request merged**